### ISSUE TITLE
Repeated gradient backgrounds are misaligned

### ISSUE THREAD

**Issue by aahancoc** [Kozea/WeasyPrint#790]
I'm having problems with repeated gradients not quite lining up with their parent containers. 

For instance, if I have a 144pt tall div, and I want a linear gradient (set `to bottom`) gradient to repeat 10 times, I'd set `background-size` to `100% 14.4pt`. This should create 10 perfectly spaced lines, starting from the top of the container. However, when Weasyprint renders it the spaces are just too large, causing it to look misaligned.

However, if I had a 150pt tall div and I used 15pt at the spacing, it works perfectly. Ditto with 125pt and 12.5pt spacing. However, 2in and 0.2in fails just like 144pt and 14.4pt. It appears that the background size is being rounded to some fraction of a point, which builds up and causes misalignments. 

I'm running on Windows 10 using Python 3.6 and Weasyprint 44.

Here's the test case file:
```html
<!doctype html>
<html>
<head>
    <meta charset="utf-8">
    <style>
        div {
            border: 0.1in solid black;
            background: linear-gradient(to bottom, red, red 0.02in, white 0.02in, white);
            margin: 0.1in;
        }
        .ex1 {width: 144pt; height: 144pt; background-size: 100% 14.4pt;}
        .ex2 {width: 150pt; height: 150pt; background-size: 100% 15pt;}
        .ex3 {width: 125pt; height: 125pt; background-size: 100% 12.5pt;}
        .ex4 {width: 2in; height: 2in; background-size: 100% 0.2in;}
    </style>
</head>
<body>
    <div class="ex1"> <!-- empty --> </div>
    <div class="ex2"> <!-- empty --> </div>
    <div class="ex3"> <!-- empty --> </div>
    <div class="ex4"> <!-- empty --> </div>
</body>
</html>
```

[And here's the output.](https://github.com/Kozea/WeasyPrint/files/2834017/test.pdf)

**Comment by liZe** [comment]
It's because we use patterns and don't repeat images and gradients manually.

I first thought that it was because we use a [`PDFSurface`](https://github.com/Kozea/WeasyPrint/blob/bdd45475b6e3363e8062fe700dea3d574563721f/weasyprint/draw.py#L468) to draw our patterns, and that `PDFSurfaces` have a size rounded to points. But the problem happens with a `RecordingSurface` too, so…

We can manually repeat images. I've tested and it works, it's not perfect (we still have small rounding problems, maybe caused by the PDF reader) but it's much better.

That would also fix #238.

**Comment by liZe** [comment]
This problem is fixed with version 53.0.
